pkg/pkg/local.go

```diff
@@ -0,0 +1,129 @@
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package pkg
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/ksonnet/ksonnet/pkg/parts"
+
+	"github.com/ksonnet/ksonnet/pkg/app"
+	"github.com/ksonnet/ksonnet/pkg/prototype"
+	"github.com/pkg/errors"
+	"github.com/spf13/afero"
+)
+
+// Local is a package based on vendored contents.
+type Local struct {
+	a            app.App
+	name         string
+	registryName string
+	config       *parts.Spec
+}
+
+var _ Package = (*Local)(nil)
+
+// NewLocal creates an instance of Local.
+func NewLocal(a app.App, name, registryName string) (*Local, error) {
+	partsPath := filepath.Join(a.VendorPath(), registryName, name, "parts.yaml")
+	b, err := afero.ReadFile(a.Fs(), partsPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading package configuration")
+	}
+
+	config, err := parts.Unmarshal(b)
+	if err != nil {
+		return nil, errors.Wrap(err, "unmarshalling package configuration")
+	}
+
+	return &Local{
+		a:            a,
+		name:         name,
+		registryName: registryName,
+		config:       config,
+	}, nil
+}
+
+// Name returns the name for the package.
+func (l *Local) Name() string {
+	return l.name
+}
+
+// RegistryName returns the registry name for the package.
+func (l *Local) RegistryName() string {
+	return l.registryName
+}
+
+// Description returns the description for the package. The description
+// is ready from the package's parts.yaml.
+func (l *Local) Description() string {
+	return l.config.Description
+}
+
+// IsInstalled returns true if the package is installed. a package is installed if it
+// has a libraries entry in app.yaml.
+func (l *Local) IsInstalled() (bool, error) {
+	// TODO: this code is duplicated in helm.go
+	libs, err := l.a.Libraries()
+	if err != nil {
+		return false, errors.Wrapf(err, "checking if package %q is installed", l.name)
+	}
+
+	_, isInstalled := libs[l.name]
+	return isInstalled, nil
+}
+
+// Prototypes returns prototypes for this package. Prototypes are defined in the
+// package's `prototypes` directory.
+func (l *Local) Prototypes() (prototype.Prototypes, error) {
+	var prototypes prototype.Prototypes
+
+	protoPath := filepath.Join(l.a.VendorPath(), l.registryName, l.name, "prototypes")
+	exists, err := afero.DirExists(l.a.Fs(), protoPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return prototypes, nil
+	}
+
+	err = afero.Walk(l.a.Fs(), protoPath, func(path string, fi os.FileInfo, err error) error {
+		if fi.IsDir() || filepath.Ext(path) != ".jsonnet" {
```

Comment:
Should we support `.libsonnet`? Are these files meant to be imported?

Reply:
prototypes have a `.jsonnet` extension by convention.